pow/src/pow.cpp

```diff
@@ -65,8 +59,53 @@ bool checkPoW(const std::string& nonce, const std::string& timestamp,
     uint64_t denominator = difficulty * lenPlusInnerFrac;
     uint64_t targetNum = std::numeric_limits<uint64_t>::max() / denominator;
 
-    uint64Bytes target;
     u64ToU8Array(targetNum, target);
+    return true;
+}
+
+bool checkPoW(const std::string& nonce, const std::string& timestamp,
+              const std::string& ttl, const std::string& recipient,
+              const std::string& data, std::string& messageHash,
+              const std::vector<pow_difficulty_t>& difficultyHistory) {
+
+    uint64_t timestampLong;
+    try {
+        timestampLong = std::stoull(timestamp);
+    } catch (...) {
+        // Should never happen, checked previously
+        return false;
+    }
+    const auto timestampMilli = std::chrono::milliseconds(timestampLong);
+
+    int difficulty = std::numeric_limits<int>::max();
+    int mostRecentDifficulty = std::numeric_limits<int>::max();
+    std::chrono::milliseconds mostRecent(0);
+
+    for (auto& thisDifficulty : difficultyHistory) {
+        const std::chrono::milliseconds t = thisDifficulty.timestamp;
+        if (t < timestampMilli && t >= mostRecent) {
+            mostRecent = t;
+            mostRecentDifficulty = thisDifficulty.difficulty;
+        }
+        const std::chrono::milliseconds lower =
+            timestampMilli - TIMESTAMP_VARIANCE;
+        const std::chrono::milliseconds upper =
+            timestampMilli + TIMESTAMP_VARIANCE;
+
+        if (t >= lower && t <= upper) {
+            difficulty = std::min(thisDifficulty.difficulty, difficulty);
+        }
+    }
+    difficulty = std::min(mostRecentDifficulty, difficulty);
+
+    const std::string payload = timestamp + ttl + recipient + data;
```

Comment:
This will result in creating unnecessary temporary std::strings. Create one string and append with other fields?